{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from math import sqrt, log\n",
    "import sys\n",
    "\n",
    "def _is_sym(seq):\n",
    "    \"\"\"Returns True if s is symmetric (same as rev. complement)\"\"\"\n",
    "    comp = {\n",
    "        'A': 'T',\n",
    "        'T': 'A',\n",
    "        'G': 'C',\n",
    "        'C': 'G'\n",
    "    }\n",
    "    return seq == ''.join([comp[i] for i in seq][::-1])\n",
    "\n",
    "\n",
    "def _overcount(st, p):\n",
    "    ocu = 0\n",
    "    x = 0\n",
    "    while True:\n",
    "        try:\n",
    "            i = st.index(p, x)\n",
    "        except ValueError:\n",
    "            break\n",
    "        ocu += 1\n",
    "        x = i + 1\n",
    "    return ocu\n",
    "\n",
    "\n",
    "def _tercorr(st):\n",
    "    _dh = 0\n",
    "    _ds = -1.4 if _is_sym(st) else 0\n",
    "    start = st[0]\n",
    "    end = st[-1]\n",
    "\n",
    "    if start == 'G' or start == 'C':\n",
    "        _dh += 0.1\n",
    "        _ds -= 2.8\n",
    "    elif start == 'A' or start == 'T':\n",
    "        _dh += 2.3\n",
    "        _ds += 4.1\n",
    "\n",
    "    if end == 'G' or end == 'C':\n",
    "        _dh += 0.1\n",
    "        _ds -= 2.8\n",
    "    elif end == 'A' or end == 'T':\n",
    "        _dh += 2.3\n",
    "        _ds += 4.1\n",
    "    return _dh, _ds\n",
    "\n",
    "def temp(s, DNA_c=5000.0, Na_c=10.0, Mg_c=20.0, dNTPs_c=10.0, uncorrected=False):\n",
    "    '''\n",
    "    Returns the DNA/DNA melting temp using nearest-neighbor thermodynamics.\n",
    "\n",
    "    This function returns better results than EMBOSS DAN because it uses updated\n",
    "    thermodynamics values and takes into account initialization parameters from\n",
    "    the work of SantaLucia (1998).\n",
    "\n",
    "    Corrects for mono- and divalent cation concentrations.\n",
    "\n",
    "    Arguments:\n",
    "    - DNA_c:   DNA concentration [nM]\n",
    "    - Na_c:    Na+ concentration [mM]\n",
    "    - Mg_c:    Mg2+ concentration [mM]\n",
    "    - dNTPs_c: dNTP concentration [mM]\n",
    "    - correction: correct for cation concentration?\n",
    "    '''\n",
    "\n",
    "    R = 1.987    # Universal gas constant (cal/(K*mol))\n",
    "    s = s.upper()\n",
    "    dh, ds = _tercorr(s)\n",
    "    k = DNA_c * 1e-9\n",
    "\n",
    "    # Adapted from Table 1 in Allawi and SantaLucia (1997).\n",
    "    # delta H (kcal/mol)\n",
    "    dh_coeffs = {\"AA\": -7.9, \"TT\": -7.9,\n",
    "                 \"AT\": -7.2,\n",
    "                 \"TA\": -7.2,\n",
    "                 \"CA\": -8.5, \"TG\": -8.5,\n",
    "                 \"GT\": -8.4, \"AC\": -8.4,\n",
    "                 \"CT\": -7.8, \"AG\": -7.8,\n",
    "                 \"GA\": -8.2, \"TC\": -8.2,\n",
    "                 \"CG\": -10.6,\n",
    "                 \"GC\": -9.8,\n",
    "                 \"GG\": -8.0, \"CC\": -8.0}\n",
    "\n",
    "    # delta S (eu)\n",
    "    ds_coeffs = {\"AA\": -22.2, \"TT\": -22.2,\n",
    "                 \"AT\": -20.4,\n",
    "                 \"TA\": -21.3,\n",
    "                 \"CA\": -22.7, \"TG\": -22.7,\n",
    "                 \"GT\": -22.4, \"AC\": -22.4,\n",
    "                 \"CT\": -21.0, \"AG\": -21.0,\n",
    "                 \"GA\": -22.2, \"TC\": -22.2,\n",
    "                 \"CG\": -27.2,\n",
    "                 \"GC\": -24.4,\n",
    "                 \"GG\": -19.9, \"CC\": -19.9}\n",
    "\n",
    "    # Multiplies the number of times each nuc pair is in the sequence by the\n",
    "    # appropriate coefficient, then returns the sum of all the pairs\n",
    "    dh = dh + \\\n",
    "        sum(_overcount(s, pair) * coeff for pair, coeff in dh_coeffs.items())\n",
    "    ds = ds + \\\n",
    "        sum(_overcount(s, pair) * coeff for pair, coeff in ds_coeffs.items())\n",
    "\n",
    "    fgc = len([filter(lambda x: x == 'G' or x == 'C', s)]) / float(len(s))\n",
    "\n",
    "    # Melting temperature\n",
    "    tm = (1000 * dh) / (ds + (R * log(k)))\n",
    "\n",
    "    if uncorrected:\n",
    "        return tm - 273.15\n",
    "\n",
    "    MNa = Na_c * 1e-3\n",
    "    MMg = Mg_c * 1e-3\n",
    "    MdNTPs = dNTPs_c * 1e-3\n",
    "\n",
    "    # Free magnesium concentration\n",
    "    Ka = 3e4  # association constant in biological buffers\n",
    "    D = (Ka * MdNTPs - Ka * MMg + 1)**2 + (4 * Ka * MMg)\n",
    "    Fmg = (-(Ka * MdNTPs - Ka * MMg + 1) + sqrt(D)) / (2 * Ka)\n",
    "\n",
    "    cation_ratio = sqrt(Fmg) / MNa if MNa > 0 else 7.0\n",
    "\n",
    "    if cation_ratio < 0.22:\n",
    "        tm = 1 / (\n",
    "            (1 / tm) +\n",
    "            ((4.29 * fgc - 3.95) * log(MNa) + 0.94 * log(MNa)**2) * 1e-5)\n",
    "    else:\n",
    "        a = 3.92\n",
    "        d = 1.42\n",
    "        g = 8.31\n",
    "        Fmg = MMg\n",
    "        if cation_ratio < 6.0:\n",
    "            a = a * (0.843 - 0.352 * sqrt(MNa) * log(MNa))\n",
    "            d = d * \\\n",
    "                (1.279 - 4.03 * log(MNa) * 1e-3 - 8.03 * log(MNa)**2 * 1e-3)\n",
    "            g = g * (0.486 - 0.258 * log(MNa) + 5.25 * log(MNa)**3 * 1e-3)\n",
    "        tm = 1 / (\n",
    "            (1 / tm) +\n",
    "            (a - 0.911 * log(Fmg) + fgc * (6.26 + d * log(Fmg)) +\n",
    "             1 / (2 * (len(s) - 1)) * (-48.2 + 52.5 * log(Fmg) +\n",
    "                                       g * log(Fmg)**2)) * 1e-5)\n",
    "\n",
    "    return tm - 273.15\n",
    "\n",
    "def kmers(datafile, kmers=16):\n",
    "    value_dict = []\n",
    "    with open(datafile, 'r') as myfile:\n",
    "        next(myfile)\n",
    "        data = myfile.read().replace('\\n', '')\n",
    "    for i in range(len(data) - kmers):\n",
    "        value_dict.append(data[i:i+kmers])\n",
    "    s = pd.Series(value_dict, name='Kmer')\n",
    "    df = pd.DataFrame(s)\n",
    "    return df\n",
    "\n",
    "def find_kmers(df, background):\n",
    "    with open(background, 'r') as myfile:\n",
    "        next(myfile)\n",
    "        data = myfile.read().replace('\\n', '')\n",
    "    \n",
    "    def find_kmer(x):\n",
    "        return data.find(x)\n",
    "    \n",
    "    def get_melting_temp(x):\n",
    "        return temp(x)\n",
    "        \n",
    "            \n",
    "    df['Location'] = df['Kmer'].map(find_kmer)\n",
    "        \n",
    "    frame = df.loc[df['Location'] != -1]\n",
    "    \n",
    "    frame['MeltTmp'] = frame['Kmer'].map(get_melting_temp)\n",
    "    \n",
    "    return frame\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76534</th>\n",
       "      <td>GGACTGGAGGGAAGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76535</th>\n",
       "      <td>GACTGGAGGGAAGCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76536</th>\n",
       "      <td>ACTGGAGGGAAGCGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76537</th>\n",
       "      <td>CTGGAGGGAAGCGTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76538</th>\n",
       "      <td>TGGAGGGAAGCGTAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76539</th>\n",
       "      <td>GGAGGGAAGCGTAGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76540</th>\n",
       "      <td>GAGGGAAGCGTAGGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76541</th>\n",
       "      <td>AGGGAAGCGTAGGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76542</th>\n",
       "      <td>GGGAAGCGTAGGAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76543</th>\n",
       "      <td>GGAAGCGTAGGAACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76544</th>\n",
       "      <td>GAAGCGTAGGAACACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76545</th>\n",
       "      <td>AAGCGTAGGAACACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76546</th>\n",
       "      <td>AGCGTAGGAACACAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76547</th>\n",
       "      <td>GCGTAGGAACACAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76548</th>\n",
       "      <td>CGTAGGAACACAGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76549</th>\n",
       "      <td>GTAGGAACACAGAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76550</th>\n",
       "      <td>TAGGAACACAGAGGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76551</th>\n",
       "      <td>AGGAACACAGAGGGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76552</th>\n",
       "      <td>GGAACACAGAGGGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76553</th>\n",
       "      <td>GAACACAGAGGGCAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76554</th>\n",
       "      <td>AACACAGAGGGCAGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76555</th>\n",
       "      <td>ACACAGAGGGCAGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76556</th>\n",
       "      <td>CACAGAGGGCAGCAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76557</th>\n",
       "      <td>ACAGAGGGCAGCAGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76558</th>\n",
       "      <td>CAGAGGGCAGCAGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76559</th>\n",
       "      <td>AGAGGGCAGCAGCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76560</th>\n",
       "      <td>GAGGGCAGCAGCCCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76561</th>\n",
       "      <td>AGGGCAGCAGCCCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76562</th>\n",
       "      <td>GGGCAGCAGCCCCACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76563</th>\n",
       "      <td>GGCAGCAGCCCCACAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76564 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Kmer\n",
       "0      NNNNNNNNNNNNNNNN\n",
       "1      NNNNNNNNNNNNNNNN\n",
       "2      NNNNNNNNNNNNNNNN\n",
       "3      NNNNNNNNNNNNNNNN\n",
       "4      NNNNNNNNNNNNNNNN\n",
       "5      NNNNNNNNNNNNNNNN\n",
       "6      NNNNNNNNNNNNNNNN\n",
       "7      NNNNNNNNNNNNNNNN\n",
       "8      NNNNNNNNNNNNNNNN\n",
       "9      NNNNNNNNNNNNNNNN\n",
       "10     NNNNNNNNNNNNNNNN\n",
       "11     NNNNNNNNNNNNNNNN\n",
       "12     NNNNNNNNNNNNNNNN\n",
       "13     NNNNNNNNNNNNNNNN\n",
       "14     NNNNNNNNNNNNNNNN\n",
       "15     NNNNNNNNNNNNNNNN\n",
       "16     NNNNNNNNNNNNNNNN\n",
       "17     NNNNNNNNNNNNNNNN\n",
       "18     NNNNNNNNNNNNNNNN\n",
       "19     NNNNNNNNNNNNNNNN\n",
       "20     NNNNNNNNNNNNNNNN\n",
       "21     NNNNNNNNNNNNNNNN\n",
       "22     NNNNNNNNNNNNNNNN\n",
       "23     NNNNNNNNNNNNNNNN\n",
       "24     NNNNNNNNNNNNNNNN\n",
       "25     NNNNNNNNNNNNNNNN\n",
       "26     NNNNNNNNNNNNNNNN\n",
       "27     NNNNNNNNNNNNNNNN\n",
       "28     NNNNNNNNNNNNNNNN\n",
       "29     NNNNNNNNNNNNNNNN\n",
       "...                 ...\n",
       "76534  GGACTGGAGGGAAGCG\n",
       "76535  GACTGGAGGGAAGCGT\n",
       "76536  ACTGGAGGGAAGCGTA\n",
       "76537  CTGGAGGGAAGCGTAG\n",
       "76538  TGGAGGGAAGCGTAGG\n",
       "76539  GGAGGGAAGCGTAGGA\n",
       "76540  GAGGGAAGCGTAGGAA\n",
       "76541  AGGGAAGCGTAGGAAC\n",
       "76542  GGGAAGCGTAGGAACA\n",
       "76543  GGAAGCGTAGGAACAC\n",
       "76544  GAAGCGTAGGAACACA\n",
       "76545  AAGCGTAGGAACACAG\n",
       "76546  AGCGTAGGAACACAGA\n",
       "76547  GCGTAGGAACACAGAG\n",
       "76548  CGTAGGAACACAGAGG\n",
       "76549  GTAGGAACACAGAGGG\n",
       "76550  TAGGAACACAGAGGGC\n",
       "76551  AGGAACACAGAGGGCA\n",
       "76552  GGAACACAGAGGGCAG\n",
       "76553  GAACACAGAGGGCAGC\n",
       "76554  AACACAGAGGGCAGCA\n",
       "76555  ACACAGAGGGCAGCAG\n",
       "76556  CACAGAGGGCAGCAGC\n",
       "76557  ACAGAGGGCAGCAGCC\n",
       "76558  CAGAGGGCAGCAGCCC\n",
       "76559  AGAGGGCAGCAGCCCC\n",
       "76560  GAGGGCAGCAGCCCCA\n",
       "76561  AGGGCAGCAGCCCCAC\n",
       "76562  GGGCAGCAGCCCCACA\n",
       "76563  GGCAGCAGCCCCACAG\n",
       "\n",
       "[76564 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_df = kmers(\"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/Chromosone16.fasta\")\n",
    "kmer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/thatcher/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:166: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/thatcher/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:170: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kmer</th>\n",
       "      <th>Location</th>\n",
       "      <th>MeltTmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1541</th>\n",
       "      <td>TTTGCTGTTCCTGCAT</td>\n",
       "      <td>1923021</td>\n",
       "      <td>66.947479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1627</th>\n",
       "      <td>TTTTCTTTGACCTCTT</td>\n",
       "      <td>1626715</td>\n",
       "      <td>59.619402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1628</th>\n",
       "      <td>TTTCTTTGACCTCTTC</td>\n",
       "      <td>1626716</td>\n",
       "      <td>60.116533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Kmer  Location    MeltTmp\n",
       "1541  TTTGCTGTTCCTGCAT   1923021  66.947479\n",
       "1627  TTTTCTTTGACCTCTT   1626715  59.619402\n",
       "1628  TTTCTTTGACCTCTTC   1626716  60.116533"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df = find_kmers(kmer_df[0:2000], \"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/GCF_000007465.2_ASM746v2_genomic.fna\")\n",
    "temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
