{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from math import sqrt, log\n",
    "import sys\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "def _is_sym(seq):\n",
    "    \"\"\"Returns True if s is symmetric (same as rev. complement)\"\"\"\n",
    "    comp = {\n",
    "        'A': 'T',\n",
    "        'T': 'A',\n",
    "        'G': 'C',\n",
    "        'C': 'G'\n",
    "    }\n",
    "    return seq == ''.join([comp[i] for i in seq][::-1])\n",
    "\n",
    "\n",
    "def _overcount(st, p):\n",
    "    ocu = 0\n",
    "    x = 0\n",
    "    while True:\n",
    "        try:\n",
    "            i = st.index(p, x)\n",
    "        except ValueError:\n",
    "            break\n",
    "        ocu += 1\n",
    "        x = i + 1\n",
    "    return ocu\n",
    "\n",
    "\n",
    "def _tercorr(st):\n",
    "    _dh = 0\n",
    "    _ds = -1.4 if _is_sym(st) else 0\n",
    "    start = st[0]\n",
    "    end = st[-1]\n",
    "\n",
    "    if start == 'G' or start == 'C':\n",
    "        _dh += 0.1\n",
    "        _ds -= 2.8\n",
    "    elif start == 'A' or start == 'T':\n",
    "        _dh += 2.3\n",
    "        _ds += 4.1\n",
    "\n",
    "    if end == 'G' or end == 'C':\n",
    "        _dh += 0.1\n",
    "        _ds -= 2.8\n",
    "    elif end == 'A' or end == 'T':\n",
    "        _dh += 2.3\n",
    "        _ds += 4.1\n",
    "    return _dh, _ds\n",
    "\n",
    "def temp(s, DNA_c=5000.0, Na_c=10.0, Mg_c=20.0, dNTPs_c=10.0, uncorrected=False):\n",
    "    '''\n",
    "    Returns the DNA/DNA melting temp using nearest-neighbor thermodynamics.\n",
    "\n",
    "    This function returns better results than EMBOSS DAN because it uses updated\n",
    "    thermodynamics values and takes into account initialization parameters from\n",
    "    the work of SantaLucia (1998).\n",
    "\n",
    "    Corrects for mono- and divalent cation concentrations.\n",
    "\n",
    "    Arguments:\n",
    "    - DNA_c:   DNA concentration [nM]\n",
    "    - Na_c:    Na+ concentration [mM]\n",
    "    - Mg_c:    Mg2+ concentration [mM]\n",
    "    - dNTPs_c: dNTP concentration [mM]\n",
    "    - correction: correct for cation concentration?\n",
    "    '''\n",
    "\n",
    "    R = 1.987    # Universal gas constant (cal/(K*mol))\n",
    "    s = s.upper()\n",
    "    dh, ds = _tercorr(s)\n",
    "    k = DNA_c * 1e-9\n",
    "\n",
    "    # Adapted from Table 1 in Allawi and SantaLucia (1997).\n",
    "    # delta H (kcal/mol)\n",
    "    dh_coeffs = {\"AA\": -7.9, \"TT\": -7.9,\n",
    "                 \"AT\": -7.2,\n",
    "                 \"TA\": -7.2,\n",
    "                 \"CA\": -8.5, \"TG\": -8.5,\n",
    "                 \"GT\": -8.4, \"AC\": -8.4,\n",
    "                 \"CT\": -7.8, \"AG\": -7.8,\n",
    "                 \"GA\": -8.2, \"TC\": -8.2,\n",
    "                 \"CG\": -10.6,\n",
    "                 \"GC\": -9.8,\n",
    "                 \"GG\": -8.0, \"CC\": -8.0}\n",
    "\n",
    "    # delta S (eu)\n",
    "    ds_coeffs = {\"AA\": -22.2, \"TT\": -22.2,\n",
    "                 \"AT\": -20.4,\n",
    "                 \"TA\": -21.3,\n",
    "                 \"CA\": -22.7, \"TG\": -22.7,\n",
    "                 \"GT\": -22.4, \"AC\": -22.4,\n",
    "                 \"CT\": -21.0, \"AG\": -21.0,\n",
    "                 \"GA\": -22.2, \"TC\": -22.2,\n",
    "                 \"CG\": -27.2,\n",
    "                 \"GC\": -24.4,\n",
    "                 \"GG\": -19.9, \"CC\": -19.9}\n",
    "\n",
    "    # Multiplies the number of times each nuc pair is in the sequence by the\n",
    "    # appropriate coefficient, then returns the sum of all the pairs\n",
    "    dh = dh + \\\n",
    "        sum(_overcount(s, pair) * coeff for pair, coeff in dh_coeffs.items())\n",
    "    ds = ds + \\\n",
    "        sum(_overcount(s, pair) * coeff for pair, coeff in ds_coeffs.items())\n",
    "\n",
    "    fgc = len([filter(lambda x: x == 'G' or x == 'C', s)]) / float(len(s))\n",
    "\n",
    "    # Melting temperature\n",
    "    tm = (1000 * dh) / (ds + (R * log(k)))\n",
    "\n",
    "    if uncorrected:\n",
    "        return tm - 273.15\n",
    "\n",
    "    MNa = Na_c * 1e-3\n",
    "    MMg = Mg_c * 1e-3\n",
    "    MdNTPs = dNTPs_c * 1e-3\n",
    "\n",
    "    # Free magnesium concentration\n",
    "    Ka = 3e4  # association constant in biological buffers\n",
    "    D = (Ka * MdNTPs - Ka * MMg + 1)**2 + (4 * Ka * MMg)\n",
    "    Fmg = (-(Ka * MdNTPs - Ka * MMg + 1) + sqrt(D)) / (2 * Ka)\n",
    "\n",
    "    cation_ratio = sqrt(Fmg) / MNa if MNa > 0 else 7.0\n",
    "\n",
    "    if cation_ratio < 0.22:\n",
    "        tm = 1 / (\n",
    "            (1 / tm) +\n",
    "            ((4.29 * fgc - 3.95) * log(MNa) + 0.94 * log(MNa)**2) * 1e-5)\n",
    "    else:\n",
    "        a = 3.92\n",
    "        d = 1.42\n",
    "        g = 8.31\n",
    "        Fmg = MMg\n",
    "        if cation_ratio < 6.0:\n",
    "            a = a * (0.843 - 0.352 * sqrt(MNa) * log(MNa))\n",
    "            d = d * \\\n",
    "                (1.279 - 4.03 * log(MNa) * 1e-3 - 8.03 * log(MNa)**2 * 1e-3)\n",
    "            g = g * (0.486 - 0.258 * log(MNa) + 5.25 * log(MNa)**3 * 1e-3)\n",
    "        tm = 1 / (\n",
    "            (1 / tm) +\n",
    "            (a - 0.911 * log(Fmg) + fgc * (6.26 + d * log(Fmg)) +\n",
    "             1 / (2 * (len(s) - 1)) * (-48.2 + 52.5 * log(Fmg) +\n",
    "                                       g * log(Fmg)**2)) * 1e-5)\n",
    "\n",
    "    return tm - 273.15\n",
    "\n",
    "def kmers(foreground, background, kmers=16, min_temp=0, max_temp=150):\n",
    "    \n",
    "    value_dict = {}\n",
    "    with open(foreground, 'r') as fg_file:\n",
    "        fg_first_line = fg_file.readline()\n",
    "        bg_data = fg_file.read().replace('\\n', '')\n",
    "        \n",
    "        with open(background, 'r') as bg_file:\n",
    "            bg_first_line = bg_file.readline()\n",
    "            fg_data = bg_file.read().replace('\\n', '')\n",
    "            \n",
    "            for i in range(int(len(bg_data)/30) - kmers):\n",
    "                cur_mer = bg_data[i:i+kmers]\n",
    "                if(cur_mer in fg_data):\n",
    "                    cur_temp = temp(cur_mer)\n",
    "                    if(min_temp <= cur_temp <= max_temp):\n",
    "                        value_dict[cur_mer] = [i, cur_temp]\n",
    "    df = pd.DataFrame.from_dict(value_dict, orient='index', columns=[\"Location\", \"Melting Temp\"])\n",
    "    return df, fg_first_line, bg_first_line\n",
    "\n",
    "def find_kmers(foregrounds, background, filepath):\n",
    "    for file in foregrounds:\n",
    "        kmer_df, fg, bg = kmers(file, background)\n",
    "        filename = filepath + 'fg' + fg.split(' ')[0].strip('>') + '_' + 'bg' + bg.split(' ')[0].strip('>') + '_optimal_mers.csv'\n",
    "        kmer_df.to_csv(filename)\n",
    "    return 0\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time Taken: 32.71850323677063\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "bg = \"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/GCF_000007465.2_ASM746v2_genomic.fna\"\n",
    "fgs = [\"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/Chromosone16.fasta\", \"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/Chromosone17.fasta\"]\n",
    "find_kmers(fgs, bg, \"/Users/thatcher/Documents/Graduate School/FofanovResearch/INF685/data/\")\n",
    "end = time.time()\n",
    "print(\"Time Taken: {}\".format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fgCM000678.2_bgNC_004350.2_optimal_mers.csv\n"
     ]
    }
   ],
   "source": [
    "print('fg' + fg.split(' ')[0].strip('>') + '_' + 'bg' + bg.split(' ')[0].strip('>') + '_optimal_mers.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
